// scriba-rule-engine/src/conditions/index.ts
import { Condition } from "../engine/RuleEngine";

export const defaultConditions = {
  greaterThan: (record: any, cond: Condition) => record[cond.field!] > cond.value,
  lessThan: (record: any, cond: Condition) => record[cond.field!] < cond.value,
  equals: (record: any, cond: Condition) => record[cond.field!] === cond.value,
  notEquals: (record: any, cond: Condition) => record[cond.field!] !== cond.value,
  fieldExists: (record: any, cond: Condition) => record[cond.field!] !== undefined,
  fieldNull: (record: any, cond: Condition) => record[cond.field!] == null,
  inList: (record: any, cond: Condition) => Array.isArray(cond.value) && cond.value.includes(record[cond.field!]),
  notInList: (record: any, cond: Condition) => Array.isArray(cond.value) && !cond.value.includes(record[cond.field!]),
  matchesRegex: (record: any, cond: Condition) => new RegExp(cond.value).test(record[cond.field!]),
  expression: (record: any, cond: Condition) => {
    if (!cond.expression) return true;
    const expr = cond.expression.replace(/{{(.*?)}}/g, (_, field) => {
      const value = field.split('.').reduce((obj: { [x: string]: any; }, key: string | number) => obj?.[key], record);
      return JSON.stringify(value);
    });
    return Function('"use strict"; return (' + expr + ')')();
  },
  beforeDate: (record: any, cond: Condition) => new Date(record[cond.field!]) < new Date(cond.value),
  afterDate: (record: any, cond: Condition) => new Date(record[cond.field!]) > new Date(cond.value),
  startsWith: (record: any, cond: Condition) => typeof record[cond.field!] === 'string' && record[cond.field!].startsWith(cond.value),
  endsWith: (record: any, cond: Condition) => typeof record[cond.field!] === 'string' && record[cond.field!].endsWith(cond.value),
  contains: (record: any, cond: Condition) => typeof record[cond.field!] === 'string' && record[cond.field!].includes(cond.value),
  allInList: (record: any, cond: Condition) => Array.isArray(record[cond.field!]) && Array.isArray(cond.value) && cond.value.every(v => record[cond.field!].includes(v)),
  anyInList: (record: any, cond: Condition) => Array.isArray(record[cond.field!]) && Array.isArray(cond.value) && cond.value.some(v => record[cond.field!].includes(v)),
};
